src/llm/http_llm_calculator.cc

```diff
@@ -0,0 +1,523 @@
+                if (!member->value.IsString())
+                    return false;
+                chat[member->name.GetString()] = member->value.GetString();
+            }
+        }
+
+        // model: string; required
+        it = this->doc.FindMember("model");
+        if (it != this->doc.MemberEnd()) {
+            if (!it->value.IsString())
+                return false;
+            this->model = it->value.GetString();
+        } else {
+            return false;
+        }
+
+        // frequency_penalty: float; optional - defaults to 0
+        it = this->doc.FindMember("frequency_penalty");
+        if (it != this->doc.MemberEnd()) {
+            if (!it->value.IsDouble())
+                return false;
+            this->frequencyPenalty = it->value.GetDouble();
+            if (this->frequencyPenalty < -2.0f || this->frequencyPenalty > 2.0f)
+                return false;
+        }
+
+        // max_tokens: int; optional
+        it = this->doc.FindMember("max_tokens");
+        if (it != this->doc.MemberEnd()) {
+            if (!it->value.IsInt())
+                return false;
+            this->maxTokens = it->value.GetInt();
+            if (this->maxTokens.value() <= 0)
+                return false;
+        }
+
+        // presence_penalty: float; optional - defaults to 0
+        it = this->doc.FindMember("presence_penalty");
+        if (it != this->doc.MemberEnd()) {
+            if (!it->value.IsDouble())
+                return false;
+            this->presencePenalty = it->value.GetDouble();
+            if (this->presencePenalty < -2.0f || this->presencePenalty > 2.0f)
+                return false;
+        }
+
+        // temperature: float; optional - defaults to 1
+        it = this->doc.FindMember("temperature");
+        if (it != this->doc.MemberEnd()) {
+            if (!it->value.IsDouble())
+                return false;
+            this->temperature = it->value.GetDouble();
+            if (this->temperature < 0.0f || this->temperature > 2.0f)
+                return false;
+        }
+
+        // top_p: float; optional - defaults to 1
+        it = this->doc.FindMember("top_p");
+        if (it != this->doc.MemberEnd()) {
+            if (!it->value.IsDouble())
+                return false;
+            this->topP = it->value.GetDouble();
+            if (this->topP < 0.0f || this->topP > 1.0f)
+                return false;
+        }
+
+        // logit_bias TODO
+        // logprops TODO
+        // top_logprobs TODO
+        // n TODO
+        // response_format TODO
+        // seed TODO
+        // stop TODO
+        // stream_options TODO
+        // tools TODO
+        // tool_choice TODO
+        // user TODO
+        // function_call TODO (deprecated)
+        // functions TODO (deprecated)
+        return true;
+    }
+};
+
+// TODO: To be moved to CB library.
+class TextStreamer {
+    std::shared_ptr<Tokenizer> tokenizer;
+    std::vector<int64_t> tokenCache;
+    size_t printLen;
```

Comment:
I would rename that field. In our case it's not about printing. It's more like a `lastReadIndex`, what do you think?

Reply:
It is copied from gen AI PR. It will be moved to CB lib anyway.